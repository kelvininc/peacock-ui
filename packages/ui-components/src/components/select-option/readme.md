# _<kv-select-option>_

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- With all properties (only label and value are mandatory) -->
<kv-select-option
	label="Option 1"
	value="option1"
	selected=true
	togglable=false>
</kv-select-option>
```


### React

```tsx
import React from 'react';
import { KvSelectOption } from '@kelvininc/react-ui-components';
export const KvSelectOptionExample: React.FC = () => (
  <>
	{/*-- With all properties (only label and value are mandatory) --*/}
	<KvSelectOption
		label="Option 1"
		link="option1"
		selected={false}
		togglable={true}
		>
	</KvSelectOption>
  </>
);
```



## Properties

| Property             | Attribute     | Description                                                                   | Type                                                                       | Default     |
| -------------------- | ------------- | ----------------------------------------------------------------------------- | -------------------------------------------------------------------------- | ----------- |
| `description`        | `description` | (optional) Description of the item displayed on the left                      | `string`                                                                   | `undefined` |
| `disabled`           | `disabled`    | (optional) If `true` the item is disabled                                     | `boolean`                                                                  | `false`     |
| `highlighted`        | `highlighted` | (optional) If `true` the item is highlighted                                  | `boolean`                                                                  | `false`     |
| `label` _(required)_ | `label`       | (required) The text to display on the item                                    | `string`                                                                   | `undefined` |
| `level`              | `level`       | (optional) The level depth at which the option is rendered                    | `number`                                                                   | `0`         |
| `options`            | --            | (optional) The children items of this option                                  | `{ [x: string]: ISelectOption; }`                                          | `{}`        |
| `selectable`         | `selectable`  | (optional) If `false` the item is only for presenting and cannot be selected. | `boolean`                                                                  | `true`      |
| `selected`           | `selected`    | (optional) If `true` the item is selected                                     | `boolean`                                                                  | `false`     |
| `state`              | `state`       | (optional) The toggle button state                                            | `EToggleState.Indeterminate \| EToggleState.None \| EToggleState.Selected` | `undefined` |
| `togglable`          | `togglable`   | (optional) If `true` the item is togglable                                    | `boolean`                                                                  | `false`     |
| `value` _(required)_ | `value`       | (required) The item value                                                     | `string`                                                                   | `undefined` |


## Events

| Event          | Description                              | Type                  |
| -------------- | ---------------------------------------- | --------------------- |
| `itemSelected` | Emitted when the user clicks on the item | `CustomEvent<string>` |


## Shadow Parts

| Part                      | Description                    |
| ------------------------- | ------------------------------ |
| `"checkbox"`              | The option's checkbox          |
| `"label"`                 | The option's label             |
| `"option-container"`      | The option's container         |
| `"select-option-content"` | The option's content container |


## CSS Custom Properties

| Name                                        | Description                                              |
| ------------------------------------------- | -------------------------------------------------------- |
| `--select-option-background-color`          | Select option background color.                          |
| `--select-option-background-color-hover`    | Select option background color when hovered.             |
| `--select-option-background-color-selected` | Select option background color when selected.            |
| `--select-option-flex-alignment`            | The alignment of items within the component's container. |
| `--select-option-height`                    | Select option height.                                    |
| `--select-option-label-color`               | Select option label color.                               |
| `--select-option-label-color-selected`      | Select option label color when selected.                 |
| `--select-option-transition-duration`       | Select option transition time.                           |


## Dependencies

### Used by

 - [kv-calendar-advanced-date-selector](../calendar-advanced-date-selector)
 - [kv-relative-time-picker](../relative-time-picker)
 - [kv-select-multi-options](../select-multi-options)
 - [kv-select-option](.)

### Depends on

- [kv-checkbox](../checkbox)
- [kv-select-option](.)

### Graph
```mermaid
graph TD;
  kv-select-option --> kv-select-option
  kv-checkbox --> kv-icon
  kv-calendar-advanced-date-selector --> kv-select-option
  kv-relative-time-picker --> kv-select-option
  kv-select-multi-options --> kv-select-option
  style kv-select-option fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


