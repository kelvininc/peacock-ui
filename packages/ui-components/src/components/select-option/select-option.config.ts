export const LEVEL_OFFSET_PX = 24;
