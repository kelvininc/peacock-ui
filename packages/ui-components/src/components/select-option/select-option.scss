@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --select-option-height: Select option height.
	 * @prop --select-option-transition-duration: Select option transition time.
	 * @prop --select-option-text-flex-direction: Sets how flex items are placed in the text flex box within the component's container.
	 * @prop --select-option-text-flex-alignment: Sets how flex items are aligned in the text flex box within the component's container.
	 * @prop --select-option-flex-alignment: The alignment of items within the component's container.
	 * @prop --select-option-background-color: Select option background color.
	 * @prop --select-option-background-color-selected: Select option background color when selected.
	 * @prop --select-option-background-color-highlighted: Select option background color when highlighted.
	 * @prop --select-option-background-color-hover: Select option background color when hovered.
	 * @prop --select-option-label-color: Select option label color.
	 * @prop --select-option-label-color-selected: Select option label color when selected.
	 * @prop --select-option-label-color-highlighted: Select option label color when highlighted.
	 * @prop --select-option-label-color-hover: Select option label color when hover.
	 * @prop --select-option-description-color: Select option description color.
	 * @prop --select-option-description-color-selected: Select option description color when selected.
	 * @prop --select-option-description-color-highlighted: Select option description color when highlighted.
	 * @prop --select-option-description-color-hover: Select option description color when hover.
	 * @prop --select-option-icon-size: Select option icon size in pixels.
	 */

	--select-option-height: 32px;
	--select-option-transition-duration: 100ms;
	--select-option-text-flex-direction: row;
	--select-option-text-flex-alignment: center;
	--select-option-flex-alignment: center;

	--select-option-background-color: #{kv-color('neutral-7')};
	--select-option-background-color-selected: #{kv-color('neutral-6')};
	--select-option-background-color-highlighted: #{kv-color('neutral-6')};
	--select-option-background-color-hover: #{kv-color('neutral-6')};

	--select-option-label-color: #{kv-color('neutral-4')};
	--select-option-label-color-selected: #{kv-color('neutral-1')};
	--select-option-label-color-highlighted: #{kv-color('neutral-2')};
	--select-option-label-color-hover: #{kv-color('neutral-2')};
	--select-option-description-color: #{kv-color('neutral-4')};
	--select-option-description-color-selected: #{kv-color('neutral-4')};
	--select-option-description-color-highlighted: #{kv-color('neutral-4')};
	--select-option-description-color-hover: #{kv-color('neutral-4')};
	--select-option-icon-size: 16px;
	--select-option-icon-color: #{kv-color('neutral-0')};
}

.select-option {
	height: var(--select-option-height);
	display: flex;
	flex-direction: row;
	align-items: var(--select-option-flex-alignment);
	user-select: none;
	cursor: pointer;
	background-color: var(--select-option-background-color);
	transition: background-color var(--select-option-transition-duration) linear;
	padding-right: $spacing-4x;
	padding-left: calc(var(--level-padding-offset) + #{$spacing-4x});

	.text-container {
		display: flex;
		min-width: 0;
		flex: 1;
		justify-content: space-between;
		flex-direction: var(--select-option-text-flex-direction);
		align-items: var(--select-option-text-flex-alignment);
	}

	.icon-container {
		margin-right: $spacing-2x;

		kv-icon {
			--icon-width: var(--select-option-icon-size);
			--icon-height: var(--select-option-icon-size);
			--icon-color: var(--select-option-icon-color);
		}
	}

	.item-label {
		@include ellipsis();
		@include kv-font-span-regular;
		transition: font-weight var(--select-option-transition-duration) linear;
		color: var(--select-option-label-color);
	}

	.item-description {
		@include kv-font-label-small-regular;
		color: var(--select-option-description-color);
	}

	&--selected {
		background-color: var(--select-option-background-color-selected);

		.item-label {
			@include kv-font-span-semibold;
			color: var(--select-option-label-color-selected);
		}

		.item-description {
			@include kv-font-label-small-semibold;
			color: var(--select-option-description-color-selected);
		}
	}

	&:hover {
		background-color: var(--select-option-background-color-hover);

		.item-label {
			color: var(--select-option-label-color-hover);
		}

		.item-description {
			color: var(--select-option-description-color-hover);
		}
	}

	&--highlighted {
		background-color: var(--select-option-background-color-highlighted);

		.item-label {
			@include kv-font-span-semibold;
			color: var(--select-option-label-color-highlighted);
		}

		.item-description {
			@include kv-font-label-small-semibold;
			color: var(--select-option-description-color-highlighted);
		}
	}

	&--disabled {
		opacity: 0.5;
		pointer-events: none;
	}

	&:not(.select-option--selectable) {
		pointer-events: none;
		user-select: none;
	}

	&--parent {
		.item-label {
			@include kv-font-label-xsmall-uppercase-semibold;
			color: kv-color('neutral-2');
		}

		.text-container::after {
			content: '';
			display: flex;
			flex: 1;
			height: 1px;
			background: kv-color('neutral-6');
			margin-left: $spacing-2x;
		}
	}

	kv-checkbox {
		--checkbox-icon-size: 12px;
		margin-right: $spacing-2x;
	}
}
