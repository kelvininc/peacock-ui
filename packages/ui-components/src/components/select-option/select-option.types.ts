import { EventEmitter } from '@stencil/core';
import { EIconName, EOtherIconName } from '../icon/icon.types';
import { ICustomCss } from '../../types';
import { HostAttributes } from '@stencil/core/internal';

export enum EToggleState {
	Selected = 'selected',
	Indeterminate = 'indeterminate',
	None = 'none'
}

export interface ISelectOption extends ICustomCss {
	/** (required) The text to display on the item */
	label: string;
	/** (required) The item value */
	value: string;
	/** (optional) Icon of the item displayed on the left */
	icon?: EIconName | EOtherIconName;
	/** (optional) Description of the item displayed on the right */
	description?: string;
	/** (optional) If `true` the item is disabled */
	disabled?: boolean;
	/** (optional) If `true` the item is selected */
	selected?: boolean;
	/** (optional) If `true` the item is highlighted */
	highlighted?: boolean;
	/** (optional) If `true` the item is togglable */
	togglable?: boolean;
	/** (optional) If `false` the item is only for presenting and cannot be selected. */
	selectable?: boolean;
	/** (optional) If `true` the item is presented as a list heading. Default: `false`*/
	heading?: boolean;
	/** (optional) The level depth at which the option is rendered */
	level?: number;
	/** (optional) The toggle button state */
	state?: EToggleState;
	/** (optional) Additional style to apply for custom CSS. */
	customStyle?: HostAttributes['style'];
	/** (optional) If true, a dirty dot indicator will be added to left side of the option's text. */
	isDirty?: boolean;
}

export interface ISelectOptionEvents {
	/** Emitted when the user clicks on the item */
	itemSelected: EventEmitter<string>;
}
