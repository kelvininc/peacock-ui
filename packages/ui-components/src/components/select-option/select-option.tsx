import { Component, Element, Host, h, Prop, EventEmitter, Event } from '@stencil/core';
import { ISelectOption, ISelectOptionEvents } from './select-option.types';
import { isNil } from 'lodash';

@Component({
	tag: 'kv-select-option',
	styleUrl: 'select-option.scss',
	shadow: true
})
export class KvSelectOption implements ISelectOption, ISelectOptionEvents {
	@Element() el!: HTMLKvSelectOptionElement;

	/** @inheritdoc */
	@Prop({ reflect: true }) label!: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) value!: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) selected?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) togglable?: boolean = false;

	/** @inheritdoc */
	@Event() itemSelected: EventEmitter<string>;

	private onItemClick = () => {
		if (!this.disabled) {
			this.itemSelected.emit(this.value);
		}
	};

	private get hasBottomSlot() {
		return !isNil(this.el.querySelector('[slot="text-bottom-slot"]'));
	}

	render() {
		return (
			<Host>
				<div
					class={{
						'select-option': true,
						'selected': this.selected,
						'disabled': this.disabled,
						'has-bottom-slot': this.hasBottomSlot
					}}
					onClick={this.onItemClick}
				>
					<div class="label-wrapper">
						{this.togglable && <kv-checkbox checked={this.selected} />}
						<div class="text-container">
							<div class="item-label">{this.label}</div>
							<slot name="text-bottom-slot"></slot>
						</div>
					</div>
					<slot name="text-right-slot"></slot>
				</div>
			</Host>
		);
	}
}
