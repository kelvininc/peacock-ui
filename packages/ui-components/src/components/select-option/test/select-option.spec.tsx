import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KvSelectOption } from '../select-option';
import { h } from '@stencil/core';

describe('KvSelectOption (unit tests)', () => {
	let page: SpecPage;
	let component: KvSelectOption;

	beforeEach(async () => {
		page = await newSpecPage({
			components: [KvSelectOption],
			template: () => <kv-select-option label="Option 1" value="option1"></kv-select-option>
		});
		component = page.rootInstance;
	});

	describe('when the component loads', () => {
		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the label', () => {
			expect(component.label).toEqual('Option 1');
		});

		it('should set the option value', () => {
			expect(component.value).toEqual('option1');
		});

		it('should set the selected status', () => {
			expect(component.selected).toEqual(false);
		});

		it('should set the togglable status', () => {
			expect(component.togglable).toEqual(false);
		});
	});
});
