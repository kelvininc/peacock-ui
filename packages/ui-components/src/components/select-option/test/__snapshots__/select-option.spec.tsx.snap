// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KvSelectOption (unit tests) when rendering with bottom slot flag \`true\` should match the snapshot 1`] = `
<kv-select-option custom-class="" label="Option 1" level="0" selectable="" value="option1">
  <mock:shadow-root>
    <div part="select-option-content">
      <div class="select-option select-option--selectable" part="option-container" style="--level-padding-offset: 0px;">
        <div class="text-container">
          <div class="item-label" part="label">
            Option 1
          </div>
        </div>
      </div>
      <slot></slot>
    </div>
  </mock:shadow-root>
</kv-select-option>
`;

exports[`KvSelectOption (unit tests) when rendering with default props when the component loads should match the snapshot 1`] = `
<kv-select-option custom-class="" label="Option 1" level="0" selectable="" value="option1">
  <mock:shadow-root>
    <div part="select-option-content">
      <div class="select-option select-option--selectable" part="option-container" style="--level-padding-offset: 0px;">
        <div class="text-container">
          <div class="item-label" part="label">
            Option 1
          </div>
        </div>
      </div>
      <slot></slot>
    </div>
  </mock:shadow-root>
</kv-select-option>
`;
